bridges/otelzerolog/hook.go

```diff
@@ -0,0 +1,112 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Package otelzerolog provides a SeverityHook, a zerolog.Hook implementation that
+// can be used to bridge between the [zerolog] API and [OpenTelemetry].
+// [zerolog]: https://github.com/rs/zerolog
+// [OpenTelemetry]: https://opentelemetry.io/docs/concepts/signals/logs/
+package otelzerolog // import "go.opentelemetry.io/contrib/bridges/otelzerolog"
+
+import (
+	"github.com/rs/zerolog"
+
+	"go.opentelemetry.io/otel/log"
+	"go.opentelemetry.io/otel/log/global"
+)
+
+type config struct {
+	provider  log.LoggerProvider
+	version   string
+	schemaURL string
+}
+
+func newConfig(options []Option) config {
+	var c config
+	for _, opt := range options {
+		c = opt.apply(c)
+	}
+
+	if c.provider == nil {
+		c.provider = global.GetLoggerProvider()
+	}
+	return c
+}
+
+func (c config) logger(name string) log.Logger {
+	var opts []log.LoggerOption
+	if c.version != "" {
+		opts = append(opts, log.WithInstrumentationVersion(c.version))
+	}
+	if c.schemaURL != "" {
+		opts = append(opts, log.WithSchemaURL(c.schemaURL))
+	}
+	return c.provider.Logger(name, opts...)
+}
+
+// Option configures a SeverityHook.
+type Option interface {
+	apply(config) config
+}
+type optFunc func(config) config
+
+func (f optFunc) apply(c config) config { return f(c) }
+
+// WithVersion returns an [Option] that configures the version of the
+// [log.Logger] used by a [SeverityHook]. The version should be the version of the
+// package that is being logged.
+func WithVersion(version string) Option {
+	return optFunc(func(c config) config {
+		c.version = version
+		return c
+	})
+}
+
+// WithSchemaURL returns an [Option] that configures the semantic convention
+// schema URL of the [log.Logger] used by a [SeverityHook]. The schemaURL should be
+// the schema URL for the semantic conventions used in log records.
+func WithSchemaURL(schemaURL string) Option {
+	return optFunc(func(c config) config {
+		c.schemaURL = schemaURL
+		return c
+	})
+}
+
+// WithLoggerProvider returns an [Option] that configures [log.LoggerProvider]
+// used by a [SeverityHook].
+//
+// By default if this Option is not provided, the SeverityHook will use the global
+// LoggerProvider.
+func WithLoggerProvider(provider log.LoggerProvider) Option {
+	return optFunc(func(c config) config {
+		c.provider = provider
+		return c
+	})
+}
+
+// SeverityHook is a [zerolog.Hook] that sends all logging records it receives to
+// OpenTelemetry. See package documentation for how conversions are made.
+type SeverityHook struct {
+	logger log.Logger
+}
+
+// NewSeverityHook returns a new [SeverityHook] to be used as a [Zerolog.Hook].
+// If [WithLoggerProvider] is not provided, the returned SeverityHook will use the
+// global LoggerProvider.
+func NewSeverityHook(name string, options ...Option) *SeverityHook {
```

Comment:
This is lacking a test that would create a new severity hook and assign it as a zerolog's hook. Even if it would do nothing yet, that would ensure the interface is properly implemented (and would be the boilerplate for future tests).

```golang
hook := NewSeverityHook("test")
log.Hook(hook)
```

Reply:
I have added the test, hope it looks good? @dmathieu

Reply:
It does look great, but it's not checking that the interface is properly implemented (if zerolog adds a new method to the interface in a new release, no test would fail).

I'm actually seeing this isn't being tested explicitly by other bridges. So maybe we should just setup an `example_test.go` file (we will need one at some point anyway)?
https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/bridges/otellogrus/example_test.go